distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/cli/action/RemovePluginAction.java

```diff
@@ -34,83 +32,80 @@
 /**
  * An action for the plugin CLI to remove plugins from Elasticsearch.
  */
-class RemovePluginAction {
-
-    // exit codes for remove
-    /** A plugin cannot be removed because it is extended by another plugin. */
-    static final int PLUGIN_STILL_USED = 11;
+public class RemovePluginAction {
 
     private final Terminal terminal;
     private final Environment env;
     private final boolean purge;
 
+    /**
+     * Categories the potential problems that {@link #checkRemovePlugins(List)} can find. Useful
+     * for generating an exit code.
+     */
+    public enum RemovePluginProblem {
+        NOT_FOUND,
+        STILL_USED,
+        BIN_FILE_NOT_DIRECTORY
+    }
+
     /**
      * Creates a new action.
      *
      * @param terminal   the terminal to use for input/output
      * @param env        the environment for the local node
      * @param purge      if true, plugin configuration files will be removed but otherwise preserved
      */
-    RemovePluginAction(Terminal terminal, Environment env, boolean purge) {
+    public RemovePluginAction(Terminal terminal, Environment env, boolean purge) {
         this.terminal = terminal;
         this.env = env;
         this.purge = purge;
     }
 
     /**
-     * Remove the plugin specified by {@code pluginName}.
-     *
-     * @param plugins    the IDs of the plugins to remove
-     * @throws IOException   if any I/O exception occurs while performing a file operation
-     * @throws UserException if plugins is null or empty
-     * @throws UserException if plugin directory does not exist
-     * @throws UserException if the plugin bin directory is not a directory
+     * Looks for problems that would prevent the specified plugins from being removed.
+     * @param plugins the plugins to check
+     * @return {@code null} if there are no problems, or a {@link Tuple} that indicates the type of problem,
+     * and a descriptive message.
+     * @throws IOException if a problem occurs loading the plugins that are currently installed.
      */
-    void execute(List<PluginDescriptor> plugins) throws IOException, UserException {
+    public Tuple<RemovePluginProblem, String> checkRemovePlugins(List<PluginDescriptor> plugins) throws IOException {
```

Comment:
What's the intention behind refactoring this to return a "problem" vs just throwing an exception? Couldn't we just as well catch the exception when calling this from bootstrap code?

Reply:
Sorry for the barrage of back and for there, this bit just seems like an odd pattern, at least in Java-land where exceptions are used to indicate errors. This seems more like Go where a method can return multiple things, including an optional error :smile:.

Reply:
So I was trying to keep exit codes and wotnot out of the Action classes and restrict them to the Command classes, trying to keep their concerns separate. But it's tricky because we have so many exit codes (really? Does anyone ever check those?). I was also trying to make it...well...less like call-a-method-and-maybe-it-will-explode.

I'm not sure what I think about this now. I'd like @rjernst's take on it.

Reply:
I think continuing to use exceptions for errors is best.